icu4j/main/classes/core/src/com/ibm/icu/util/ULocale.java

```diff
@@ -1203,6 +1205,328 @@ public static String getKeywordValue(String localeID, String keywordName) {
+
+                if (replacedLanguage.equals(UNDEFINED_LANGUAGE)) {
+                    replacedLanguage = language;
+                }
+                String replacedScript = null;
+                String replacedRegion = null;
+                String replacedVariant = null;
+                String replacedExtensions = null;
+                int consumed = replacementFields[0].length() + 1;
+                while (replacementFields.length > index) {
+                    String field = replacementFields[index];
+                    int len = field.length();
+                    if (1 == len) {
+                        replacedExtensions = replacement.substring(consumed);
+                        break;
+                    } else if (len >= 2 && len <= 3) {
+                        assert replacedRegion == null;
+                        replacedRegion = field;
+                    } else if (len >= 5 && len <= 8) {
+                        assert replacedVariant == null;
+                        replacedVariant = field;
+                    } else if (len == 4) {
+                        if (field.charAt(0) >= '0' && field.charAt(0) <= '9') {
+                            assert replacedVariant == null;
+                            replacedVariant = field;
+                        } else {
+                            assert replacedScript == null;
+                            replacedScript = field;
+                        }
+                    }
+                    index++;
+                    consumed += len + 1;
+                }
+
+                replacedScript = deleteOrReplace(script, null, replacedScript);
+                replacedRegion = deleteOrReplace(region, searchRegion, replacedRegion);
+                replacedVariant = deleteOrReplace(searchVariant, searchVariant, replacedVariant);
+
+                if (same(this.language, replacedLanguage) &&
+                    same(this.script, replacedScript) &&
+                    same(this.region, replacedRegion) &&
+                    same(searchVariant, replacedRegion) &&
+                    replacedExtensions == null) {
+                    variantIndex++;
+                    // Replacement produce no changes on search.
+                    // For example, apply pa_IN=> pa_Guru_IN on pa_Guru_IN.
+                    continue;
+                }
+                this.language = replacedLanguage;
+                this.script = replacedScript;
+                this.region = replacedRegion;
+                if (searchVariant != null && !searchVariant.isEmpty()) {
+                    if (replacedVariant != null && !replacedVariant.isEmpty()) {
+                        this.variants.set(variantIndex, replacedVariant);
+                    } else {
+                        this.variants.remove(variantIndex);
+                    }
+                }
+                if (replacedExtensions != null) {
+                    this.extensions = replacedExtensions;
```

Comment:
This clobbers all existing extensions... What are the possible values here? Is it just for -u-va-posix? Should we try to "mix that in"? TODO?

Reply:
We currently have four entries which have extension in langaugeAlias :
```
            i_default{
                reason{"legacy"}
                replacement{"en_x_i_default"}
            }
            i_enochian{
                reason{"legacy"}
                replacement{"und_x_i_enochian"}
            }
            i_mingo{
                reason{"legacy"}
                replacement{"see_x_i_mingo"}
            }
            zh_min{
                reason{"legacy"}
                replacement{"nan_x_zh_min"}
            }
```

The type (which is the key in the resources) are not legal under uts35 unicode identifier syntax, so the only time we will get into problem is if we have something like
zh-min-u-ca-chinese
or
i-enochian-u-kf-upper

@macchiati 
My understanding is according to Mark's new spec (https://docs.google.com/document/d/1x14W-jBfra3Y-Lg-UfnippIMmU29VISI8dSAAUCcRhE/edit#) , these entries are considered as part of "BCP47 LegacyRules" But his draft spec stated:
"If the source is identical to one of the types in the BCP47 LegacyRules, replace the entire source by the replacement value.
"
It seems ULocale, before the code reach this point, already change 
zh-min to nan@x=zh-min
and change
zh-min-u-ca-chinese to nan@x=zh-min-u-ca-chinese

I am not sure we should do a "mix that in" unless Mark's spec said we should do that.

Reply:
ULocale.toString() uses the old `@` syntax for compatibility.
ULocale.toLanguageTag() uses `-x-...`.
Please submit a ticket and add a TODO here to revisit this.

Reply:
ok https://unicode-org.atlassian.net/browse/ICU-21292